src/metabase/db/util.clj

```diff
@@ -66,3 +68,71 @@
+     ~@body
+     (catch ExceptionInfo e#
+       ;; The underlying exception thrown by the driver is database specific and opaque, so we treat any exception as a
+       ;; possible database conflict due to a concurrent insert. If we want to be more conservative, we would need
+       ;; a per-driver or driver agnostic way to test the exception.
+       ~@body)))
+
+(defn select-or-insert!
+  "Return a database record if it exists, otherwise create it.
+
+   This is more general than using `UPSERT`, `MERGE` or `INSERT .. ON CONFLICT`, and it also allows one to avoid
+   calculating initial values that may be expensive, or require side effects.
+
+   In the case where there is an underlying db constraint to prevent duplicates, this method takes care of handling
+   rejection from the database due to a concurrent insert, and will retry a single time to pick up the existing row.
+   This may result in `insert-fn` being called a second time.
+
+   In the case where there is no underlying db constraint, concurrent calls may still result in duplicates.
+   To prevent this in a database agnostic way, during an existing non-serializable transaction, would be non-trivial."
+  {:style/indent 2}
+  [model select-map insert-fn]
+  (with-conflict-retry
+   (let [select-kvs (mapcat identity select-map)
+         validate   (fn [updated#]
+                      (assert (= select-map (merge select-map (select-keys updated# (keys select-map))))
+                              "This should not be used to change any of the identifying values")
+                      ;; For convenience, we allow the insert-fn to omit fields in the search-map
+                      (merge updated# select-map))]
+     (or (apply t2/select-one model select-kvs)
+         (t2/insert-returning-instance! model (validate (insert-fn)))))))
+
+(defn update-or-insert!
+  "Update a database record, if it exists, otherwise create it.
+
+   This is more general than using `UPSERT`, `MERGE` or `INSERT .. ON CONFLICT`, and it also allows one to avoid
+   calculating initial values that may be expensive, or require side effects.
+
+   In the case where there is an underlying db constraint to prevent duplicates, this method takes care of handling
+   rejection from the database due to a concurrent insert, and will retry a single time to pick up the existing row.
+   This may result in `insert-fn` being called a second time.
+
+   In the case where there is no underlying db constraint, concurrent calls may still result in duplicates.
+   To prevent this in a database agnostic way, during an existing non-serializable transaction, would be non-trivial."
+  {:style/indent 2}
+  [model select-map update-fn]
+  (with-conflict-retry
+   (let [select-kvs (mapcat identity select-map)
+         pks        (t2/primary-keys model)
+         _          (assert (= 1 (count pks)) "This helper does not currently support compound keys")
+         pk-key     (keyword (first pks))
+         validate   (fn [updated]
```

Comment:
`validate` is a misleading name because its main operation is merging with a select map. The assertion is secondary imo, although that behaviour could be documented in the docstring.
```suggestion
         merge-with-select-map   (fn [updated]
```

Reply:
I struggled with this name, and eventually went with my bias for short names.

Re: primary / secondary, I wanted to highlight validation because `merge` doesn't sound like something that could throw an exception. I think you're right though that it should focus on the transformation, perhaps a comment about the capability to throw is enough.

Reply:
Ok so I thought about this, and I think I might have found a naming loophole.

We need to remember to always call this function on the result of `update-fn`, so I'm inclined to fuse them together, and I think this may ironically make it easier to name.

```clojure
;; wrap update-fn to assert we preserve our identifiers, and make them explicit
(fn [existing]
  (let [updates (update-fn existing)]
    (assert (empty? (u/conflicting-keys select-map updates))
    (merge select-map updates))
```

I'm inclined to just shadow `update-fn`, or something obviously related like `safe-update-fn`.

Reply:
Nice. Yeah shadowing `update-fn` works, or naming it `update-fn'` even